`security.wrappers`: init

Samasaur1 opened the pull request:
This PR adds `security.wrappers` to nix-darwin, closing #618. It's heavily modeled after how NixOS implements the same option. On Darwin, it allows setting ownership, permissions, setuid/setgid, and (not yet implemented) codesigning. I've tested this PR, and it does meet the minimal use case of providing a setuid wrapper to a given program.

However, I have some concerns that should be addressed before it is merged:
- Is the generated wrapper "secure"? The NixOS implementation took a lot of steps to ensure security, some of which I could not replicate on Darwin. For example, multiple lists of unsecure environment variables — loaded from other derivations, not defined as constants — were combined into the wrapper. I've copied what the env vars were at the time of writing, but this isn't great for maintenance
- Supporting codesigning. I wasn't exactly clear on what the goal was here, only that it was mentioned in #618. I left a stub in this implementation which should hopefully allow implementing codesigning without a huge amount of reworking.
- Am I correct that setting capabilities is not a thing on Darwin?
- NixOS does not let you use setcap and setuid on the same wrapper. Although I didn't implement setcap at all, I do allow codesigning and setuid/setgid on the same wrapper. Is there a reason to disallow this?

These are just the concerns that I have — if others have concerns (about the concept as a whole or about my implementation), please leave a comment.

Samasaur1 committed:
Potential first version of security.wrappers

Samasaur1 committed:
Actually wrap with C program

Samasaur1 committed:
Trim C wrapper to work on Darwin

Samasaur1 committed:
include libs

Samasaur1 committed:
unset env vars

Samasaur1 commented:
Is it possible that codesigning is desired so that entitlements can be applied to binaries?

Samasaur1 committed:
Update examples

Samasaur1 committed:
Remove codesigning support; leave stubs in place

Samasaur1 commented:
Okay, I have removed codesigning from this PR, so it now only supports setuid/setgid. I've left the commented-out codesigning code in place for now, but I can remove it if desired.

I'm marking this PR as ready for review, and hopefully reviewers can look at `wrapper.c`/`wrapper.nix` in this PR versus [wrapper.c](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/security/wrappers/wrapper.c) and [wrapper.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/security/wrappers/wrapper.nix) in nixpkgs, with a focus on the security of the wrappers.

Samasaur1 committed:
Create wrapper directory at boot

Without this change, the wrapper directory would be removed on reboot
and would not be recreated until the next time the activation script ran

Samasaur1 committed:
Fix tests

Samasaur1 committed:
Replace Karabiner-specific /run/wrappers/bin implementation

Samasaur1 commented:
Pinging @Enzime for the change to Karabiner

Samasaur1 committed:
Remove mdDoc as it has been deprecated

Enzime reviewed (commented):
The Karabiner change looks fine to me 👍

Samasaur1 committed:
Require services.activate-system.enable for security.wrappers

`/run` is wiped on boot, and services.activate-system restores it.
Technically a user could use security.wrappers without
services.activate-system so long as they were willing to rerun the
activation script after every reboot, but some services (such as
Karabiner) add programs to security.wrappers and expect that they exist
on or shortly after boot

emilazy commented on modules/security/wrappers/wrapper.c:
```diff
@@ -0,0 +1,128 @@
+    abort();
+}
+
+// These are environment variable aliases for glibc tunables.
+// This list shouldn't grow further, since this is a legacy mechanism.
+// Any future tunables are expected to only be accessible through GLIBC_TUNABLES.
+//
+// They are not included in the glibc-provided UNSECURE_ENVVARS list,
+// since any SUID executable ignores them. This wrapper also serves
+// executables that are merely granted ambient capabilities, rather than
+// being SUID, and hence don't run in secure mode. We'd like them to
+// defend those in depth as well, so we clear these explicitly.
+//
+// Except for MALLOC_CHECK_ (which is marked SXID_ERASE), these are all
+// marked SXID_IGNORE (ignored in secure mode), so even the glibc version
+// of this wrapper would leave them intact.
+#define UNSECURE_ENVVARS_TUNABLES \
+    "MALLOC_CHECK_\0" \
+    "MALLOC_TOP_PAD_\0" \
+    "MALLOC_PERTURB_\0" \
+    "MALLOC_MMAP_THRESHOLD_\0" \
+    "MALLOC_TRIM_THRESHOLD_\0" \
+    "MALLOC_MMAP_MAX_\0" \
+    "MALLOC_ARENA_MAX\0" \
+    "MALLOC_ARENA_TEST\0"
```
Not sure if any of these apply to macOS.

Samasaur1 replied:
Me neither, but I'm leaving them in to err on the side of safety

emilazy commented on modules/security/wrappers/wrapper.c:
```diff
@@ -0,0 +1,128 @@
+  "GCONV_PATH\0"							      \
+  "GETCONF_DIR\0"							      \
+  "HOSTALIASES\0"							      \
+  "LD_AUDIT\0"								      \
+  "LD_DEBUG\0"								      \
+  "LD_DEBUG_OUTPUT\0"							      \
+  "LD_DYNAMIC_WEAK\0"							      \
+  "LD_HWCAP_MASK\0"							      \
+  "LD_LIBRARY_PATH\0"							      \
+  "LD_ORIGIN_PATH\0"							      \
+  "LD_PRELOAD\0"							      \
+  "LD_PROFILE\0"							      \
+  "LD_SHOW_AUXV\0"							      \
+  "LD_USE_LOAD_BIAS\0"							      \
```
These are specific to the Linux dynamic loader interface – **meaning that their Darwin `DYLD_*` equivalents are currently not sanitized at all**. I fear that this means the current version of the wrapper is likely vulnerable.

I suspect that there are other Linux‐specific variables still in this list that I’m less confident about, and almost surely Darwin‐specific ones that we’re missing. That makes me really uneasy.

Samasaur1 replied:
I have updated the wrapper to unset the Darwin-specific env vars, although I have continued to unset these env vars anyway

emilazy commented on modules/security/wrappers/wrapper.c:
It’s possible that Darwin’s dynamic loader also does this, in which case we could sleep a lot easier about all this. It would be good if someone could investigate that.

Samasaur1 replied:
The dynamic loader does drop all the `DYLD_*` env vars that were listed in `man dyld`

emilazy commented on modules/security/wrappers/wrapper.c:
Not relevant to us, except maybe for codesigning entitlements.

Samasaur1 replied:
I've updated this comment to refer to binaries with entitlements

emilazy commented on modules/security/wrappers/wrapper.c:
This (and other glibc‐related comments) would need to be updated.

emilazy reviewed (changes requested):
So, to be upfront about my bias here: I really dislike set{u,g}id. They’re mechanisms for running processes in highly‐privileged contexts but – critically – with an environment that is almost entirely under the control of an *untrusted* context. Privileged daemons interacting with untrusted clients are already a security minefield because of the confused deputy problem, but while those get started by a service manager in a predictable and controlled environment, set{u,g}id multiply the problem by allowing the invoking process to arrange pretty much any context for the hapless privileged code they desire. Naturally, it’s extremely difficult to code to this standard, so this leads to all kinds of awful vulnerabilities.

Most of the time, you are better off setting up a privileged daemon or a restricted `sudo` rule. `sudo` itself is a set{u,g}id program, of course, but it does a lot of the job of setting up a more predictable environment for its child processes itself, and gets more security attention than the majority of set{u,g}id code out there or especially our wrappers. (Despite this, it still has a lot of vulnerabilities! systemd’s `run0`, Ian Jackson’s `userv`, or even `ssh` to localhost are fundamentally better approaches here.)

In this case, I understand the examples people have brought up are:

* `ping`: This actually works fine without `setuid`. You only need privileges for raw pings, which aren’t that common to use. A simple restricted `sudo` rule is likely a better option here if ordinary users being able to send raw pings without privilege elevation is critical.

* `mtr`: Also a good candidate for `sudo`.

* `top`/`htop`: These ones are trickier. I’m not sure about `top`, but you probably don’t want to give people unrestricted access to run `htop` as root, since it can kill processes and the like. *However,* since it’s already coded to be able to be run as setuid, I believe you can instruct `sudo` to set the real/effective UIDs and GIDs such that you don’t get any additional privileges.

I would feel much more comfortable with reviewing and merging options that try to set up locked down `sudo` rules and wrappers that call out to `sudo` for cases like these.

For what it’s worth, I think that codesigning would actually be a great use case for this kind of interface, though; being able to set entitlements on executables would be tighter‐scoped than full set{u,g}id wrappers in the same way that Linux capabilities are, and unlock functionality we really have no other way of providing at the moment.

With all that said, support for set{u,g}id wrappers is already an unfortunate reality on NixOS, so if people are really clamouring for it then I’m not going to stop this going through. I just want to be clear that *here be dragons* and we should expect it to lead to vulnerabilities; even NixOS’s wrapper has had security issues. I wish I could say I feel fully qualified to sign off on an implementation, but I’d feel more comfortable getting a review from someone who is more intimately familiar with the fine details of Darwin’s dynamic loader. If we can’t get that then we might just want to put a big warning on the options before merging.

Samasaur1 committed:
Fix shellcheck errors

Samasaur1 committed:
security.wrappers: unset Darwin-specific env vars

Samasaur1 committed:
security.wrappers: update comment (glibc -> dyld)

Samasaur1 committed:
security.wrappers: remove assertion on services.activate-system.enable

The option was removed since it is now always on.

Samasaur1 committed:
security.wrappers: copy source programs instead of wrapping in binaries

Copying the wrapper code from NixOS and hoping that I managed to catch
all the relevant dyld environment variables was not particularly
confidence-inducing, so this commit removes the wrappers entirely and
simply copies the source programs before modifying permissions. This
means that the given source programs must safely handle being
setuid/setgid binaries and being located in `/run/wrappers/bin`.

There's now only `default.nix` in `modules/security/wrappers`, so I
could have removed the directory, but I have left it for now in
anticipation of other files potentially ending up there.

dwt commented:
How could an alternative look that used sudo rules to achieve similar goals?

For `mtr` could that be replacing the `mtr-packet` application with a shell wrapper that `exec`'s the original binary through sudo to elevate it's permissions?

That could be relatively straightforward to implement.

I am trying to gage what a simple and secure implementation could look like.
